# Case study


---





What should be presented in the case study:


Degree (strength) distribution



Two of the following characteristics:

1.  Average path length

2.  Clustering coefficient (Arushi)

3.  Small-worldness


---



Main centrality measures:

1. Degree - how connected the node is   (Arushi)

2. Closeness - how easily a node can reach other nodes

3. Betweenness - how important the node is in terms of connecting the other nodes (Arushi)

4. Neighbors' characteristics (prestige) - how important are the node's neighbors

In [ ]:
#imports
import pandas as pd
import numpy as np
import random
import networkx as nx


In [ ]:
#load Data
df_mobility = pd.read_excel(r'/content/drive/MyDrive/Network Eco I/Exploratory_Analysis_Graphs/cleanedExcel.xlsx')

In [ ]:
df_mobility = df_mobility.iloc[: , 1:]


In [ ]:
df_mobility.info()

# Network Creation


In [ ]:
#creating the networks
import networkx as nx
from datetime import datetime
i=0
j=0
k=0
latdodici=[]
lgdodici=[]

latquindici=[]
lgquindici=[]

latall=[]
lgall=[]

posdodici={}
posquindici={}
posventuno={}

autonomia_auto=200

for lat, lng, temp in zip(df_mobility['latitude_[dg]'], df_mobility['longitude_[dg]'], df_mobility['commissioning_date']):
    t1=datetime.strptime(str(temp), "%Y-%m-%d %H:%M:%S")
    if t1.year<2012:
        i=i+1
        latdodici.append(lat)
        lgdodici.append(lng)
        posdodici[i]=(lng,lat)
    
    if t1.year<2016:
        j=j+1
        latquindici.append(lat)
        lgquindici.append(lng)
        posquindici[j]=(lng,lat)
    
    k=k+1
    latall.append(lat)
    lgall.append(lng)
    posventuno[k]=(lng,lat)

# print(latquindici[50])

from math import sin,cos,sqrt,atan2,radians

Gdodici=np.zeros((i,i))

R = 6373.0 # planet radius

for a in range (1,i):
    for b in range (1,i):
        #print("Questa a", a, "questa ist b", b)
        if a==b:
            Gdodici[a,b]=0
        else:
            dlon=radians(lgdodici[a]-lgdodici[b])
            dlat=radians(latdodici[a]-latdodici[b])
            aD = (sin(dlat/2))**2 + cos(radians(latdodici[a])) * cos(radians(latdodici[b])) * (sin(dlon/2))**2
            c = 2 * atan2(sqrt(aD), sqrt(1-aD))
            distance = R * c
            if distance<autonomia_auto: #3ßß*0.8
                Gdodici[a,b]=1

# print(Gdodici)

options = {
    'node_color': 'lavender',
    'node_size': 300,
    'width': 1,
    'arrowstyle': '->',
    'arrowsize': 1,
}
#2012: networkdodici
networktwelve=nx.Graph(Gdodici)
# adjadodici=nx.DiGraph(Gdodici)
mapping = dict(zip(np.arange(0,i), np.arange(1,i+1)))

print('Calculating 2015')
Gquindici=np.zeros((j,j))

for a in range (1,j):
    for b in range (1,j):
        #print("Questa a", a, "questa ist b", b)
        if a==b:
            Gquindici[a,b]=0
        else:
            dlon=radians(lgquindici[a]-lgquindici[b])
            dlat=radians(latquindici[a]-latquindici[b])
            aD = (sin(dlat/2))**2 + cos(radians(latquindici[a])) * cos(radians(latquindici[b])) * (sin(dlon/2))**2
            c = 2 * atan2(sqrt(aD), sqrt(1-aD))
            distance = R * c
            if distance<autonomia_auto: #3ßß*0.8
                Gquindici[a,b]=1

networkfifthteen=nx.Graph(Gquindici)
# adjadodici=nx.DiGraph(Gdodici)
mappingquindici = dict(zip(np.arange(0,j), np.arange(1,j+1)))



Calculating 2015


In [ ]:
print('Calculating 2021')
Gventuno=np.zeros((k,k))

for a in range (1,k):
    for b in range (1,k):
        #print("Questa a", a, "questa ist b", b)
        if a==b:
            Gventuno[a,b]=0
        else:
            dlon=radians(lgall[a]-lgall[b])
            dlat=radians(latall[a]-latall[b])
            aD = (sin(dlat/2))**2 + cos(radians(latall[a])) * cos(radians(latall[b])) * (sin(dlon/2))**2
            c = 2 * atan2(sqrt(aD), sqrt(1-aD))
            distance = R * c
            if distance<autonomia_auto: #3ßß*0.8
                Gventuno[a,b]=1

networktwentyone=nx.Graph(Gventuno)
# adjadodici=nx.DiGraph(Gdodici)
mappingventuno = dict(zip(np.arange(0,k), np.arange(1,k+1)))

In [ ]:
print(nx.info(networktwelve))
print(nx.info(networkfifthteen))


Graph with 421 nodes and 38405 edges
Graph with 1599 nodes and 363751 edges


# General Network Properties





TO-DO\\\\\\\\\\


Degree (strength) distribution



---


2012:  Graph with 421 nodes and 38405 edges

2015:  Graph with 1599 nodes and 363751 edges

In [ ]:
list(networktwelve.nodes)

In [ ]:
list(networktwelve.edges)


In [ ]:
list(networktwelve.adj[1])

In [ ]:
list(nx.connected_components(networktwelve))


#Network Characteristics

##  Clustering  Coefficient

---


  Estimates the average clustering coefficient of G.
The local clustering of each node in `G` is the
	fraction of triangles that actually exist over
	all possible triangles in its neighborhood.
	The average clustering coefficient of a graph
	`G` is the mean of local clusterings.

This function finds an approximate average
	clustering coefficient for G by repeating `n`
	times (defined in `trials`) the following
	experiment: choose a node at random, choose
	two of its neighbors at random, and check if
	they are connected. The approximate coefficient
	is the fraction of triangles found over the
	number of trials [1]_.


---



	


In [ ]:
#clustering coeff for every node
print(nx.clustering(networktwelve))
print(nx.clustering(networkfifthteen))

{0: 0, 1: 0.8104838709677419, 2: 0.5903755868544601, 3: 0.612877263581489, 4: 0.612877263581489, 5: 0.612877263581489, 6: 0.5981735159817352, 7: 0.703343070431678, 8: 0.7159507159507159, 9: 0.7644927536231884, 10: 0.9809523809523809, 11: 0.7383966244725738, 12: 0.7383966244725738, 13: 0.7416267942583732, 14: 0.7383966244725738, 15: 0.8063063063063063, 16: 0.8063063063063063, 17: 0.8063063063063063, 18: 0.5887005649717514, 19: 0.9287356321839081, 20: 0.9333333333333333, 21: 0.9733333333333334, 22: 0.9448275862068966, 23: 0.9373219373219374, 24: 0.9373219373219374, 25: 0.7129807692307693, 26: 0.8768939393939394, 27: 0.8331834981520327, 28: 0.8275184275184275, 29: 0.8276699029126213, 30: 0.8063063063063063, 31: 0.8063063063063063, 32: 0.875, 33: 0.7353658536585366, 34: 0.6114605171208944, 35: 0.7809388335704125, 36: 0.9161290322580645, 37: 0.9333333333333333, 38: 0.9333333333333333, 39: 0.9630541871921182, 40: 0.9333333333333333, 41: 0.9333333333333333, 42: 0.9096774193548387, 43: 0.96305

In [ ]:
def average_clustering(G, trials=1000):
  #Parameters
	#G : NetworkX graph

	#trials : integer
		#Number of trials to perform (default 1000).

	#Returns
	
	#c : float
		#Approximated average clustering coefficient.

	n = len(G)
	triangles = 0
	nodes = G.nodes()
	for i in [int(random.random() * n) for i in range(trials)]:
		nbrs = list(G[nodes[i]])
		if len(nbrs) < 2:
			continue
		u, v = random.sample(nbrs, 2)
		if u in G[v]:
			triangles += 1
	return triangles / float(trials)


In [ ]:
print(nx.average_clustering(networktwelve))


0.8955958561891517


In [ ]:
print(nx.average_clustering(networkfifthteen))


0.7955058411521819



Average Clustering 2012:    0.8955958561891517


---



Average Clustering 2015:    0.7955058411521819


In [ ]:
subgraphs = (G.subgraph(c) for c in nx.connected_components(G))

len_sub = []
cluster_sub = []
avg_path = []

for C in subgraphs:
    len_sub.append(len(C))
    cluster_sub.append(nx.average_clustering(C))
    avg_path.append(nx.average_shortest_path_length(C))
    
    
df_net_char = pd.DataFrame.from_dict({'N':len_sub,'Clustering Coeff':  cluster_sub,'Average Path': avg_path})

In [ ]:
df_net_char.sort_values(by = 'N', ascending = False)


##Small Worldness
Functions for estimating the small-world-ness of graphs.
A small world network is characterized by a small average shortest path length,
and a large clustering coefficient.



#Centrality Measures

##Degree Centrality













```
# nx.degree_centrality(G)
```




The degree centrality values are normalized by dividing by the maximum possible degree in a simple graph n-1 where n is the number of nodes in G.


---
## Betweenness centrality




```
#  nx.betweenness_centrality(G)
```


Betweeness of a node v is the sum of the fraction of all-pairs shortest 
paths that pass through 




In [ ]:
degree_centrality = nx.degree_centrality(G)
betweenness_centrality = nx.betweenness_centrality(G)

centralities = pd.DataFrame.from_dict({'DC': degree_centrality, 'BC': betweenness_centrality})
centralities.sort_values(by = 'DC', ascending = False)

In [ ]:
#sorted list of BC and DC
centralities.sort_values(by = 'BC', ascending = False)

DC        BC
141  0.471429  0.076987
142  0.226190  0.057048
173  0.426190  0.055746
409  0.185714  0.039292
410  0.185714  0.039292
..        ...       ...
104  0.173810  0.000008
43   0.069048  0.000008
39   0.069048  0.000008
10   0.035714  0.000004
0    0.000000  0.000000

[421 rows x 2 columns]